certbot-nginx/certbot_nginx/nginxparser.py

```diff
@@ -115,24 +134,111 @@ def load(_file):
+def dump(blocks, _file):
     """Dump to a file.
 
-    :param list block: The parsed tree
+    :param UnspacedList block: The parsed tree
     :param file _file: The file to dump to
     :param int indentation: The number of spaces to indent
     :rtype: NoneType
 
     """
-    return _file.write(dumps(blocks, indentation))
+    return _file.write(dumps(blocks))
+
+
+spacey = lambda x: (isinstance(x, str) and x.isspace()) or x == ''
+
+class UnspacedList(list):
+    """Wrap a list [of lists], making any whitespace entries magically invisible"""
+
+    def __init__(self, list_source):
+        # ensure our argument is not a generator, and duplicate any sublists
+        self.spaced = copy.deepcopy(list(list_source))
+
+        # Turn self into a version of the source list that has spaces removed
+        # and all sub-lists also UnspacedList()ed
+        list.__init__(self, list_source)
+        for i, entry in reversed(list(enumerate(self))):
+            if isinstance(entry, list):
+                sublist = UnspacedList(entry)
+                list.__setitem__(self, i, sublist)
+                self.spaced[i] = sublist.spaced
+            elif spacey(entry):
+                # don't delete comments
+                if "#" not in self[:i]:
+                    list.__delitem__(self, i)
+
+    def _coerce(self, inbound):
+        """
+        Coerce some inbound object to be appropriately usable in this object
+
+        :param inbound: string or None or list or UnspacedList
+        :returns: (coerced UnspacedList or string or None, spaced equivalent)
+        :rtype: tuple
+
+        """
+        if not isinstance(inbound, list):                      # str or None
+            return (inbound, inbound)
+        else:
+            if not hasattr(inbound, "spaced"):
+                inbound = UnspacedList(inbound)
+            return (inbound, inbound.spaced)
+
+
+    def insert(self, i, x):
+        item, spaced_item = self._coerce(x)
+        self.spaced.insert(i + self._spaces_before(i), spaced_item)
+        list.insert(self, i, item)
+
+    def append(self, x):
+        item, spaced_item = self._coerce(x)
+        self.spaced.append(spaced_item)
+        list.append(self, item)
+
+    def extend(self, x):
+        item, spaced_item = self._coerce(x)
+        self.spaced.extend(spaced_item)
+        list.extend(self, item)
+
+    def __add__(self, other):
+        l = copy.deepcopy(self)
+        l.extend(other)
+        return l
+
+    def __setitem__(self, i, value):
+        item, spaced_item = self._coerce(value)
+        self.spaced.__setitem__(i + self._spaces_before(i), spaced_item)
+        list.__setitem__(self, i, item)
+
+    def __delitem__(self, i):
+        self.spaced.__delitem__(i + self._spaces_before(i))
+        list.__delitem__(self, i)
+
+    def __deepcopy__(self, unused_memo):
+        l = UnspacedList(self[:])
+        l.spaced = copy.deepcopy(self.spaced)
+        return l
+
+
+    def _spaces_before(self, idx):
```

Comment:
You're right, this is buggy when used with constructions like `unsigned_list[-2] = thing`, though the symptom seems to be crashing rather than infinite looping.